components/Ripple/src/MDCStatefulRippleView.m

```diff
@@ -0,0 +1,229 @@
+  CGPoint _lastTouch;
+}
+
+@dynamic activeRippleLayer;
+
+- (instancetype)initWithFrame:(CGRect)frame {
+  self = [super initWithFrame:frame];
+  if (self) {
+    [self commonMDCStatefulRippleViewInit];
+  }
+  return self;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)coder {
+  self = [super initWithCoder:coder];
+  if (self) {
+    [self commonMDCStatefulRippleViewInit];
+  }
+  return self;
+}
+
+- (void)commonMDCStatefulRippleViewInit {
+  if (_rippleColors == nil) {
+    _rippleColors = [NSMutableDictionary dictionary];
+    UIColor *selectionColor = RippleSelectedColor();
+    _rippleColors[@(MDCRippleStateNormal)] = [UIColor colorWithWhite:0 alpha:kRippleAlpha];
+    _rippleColors[@(MDCRippleStateHighlighted)] = [UIColor colorWithWhite:0 alpha:kRippleAlpha];
+    _rippleColors[@(MDCRippleStateSelected)] =
+        [selectionColor colorWithAlphaComponent:kRippleSelectedAlpha];
+    _rippleColors[@(MDCRippleStateSelected | MDCRippleStateHighlighted)] =
+        [selectionColor colorWithAlphaComponent:kRippleAlpha];
+    _rippleColors[@(MDCRippleStateDragged)] = [UIColor colorWithWhite:0 alpha:kRippleDraggedAlpha];
+    _rippleColors[@(MDCRippleStateDragged | MDCRippleStateHighlighted)] =
+        [UIColor colorWithWhite:0 alpha:kRippleDraggedAlpha];
+    _rippleColors[@(MDCRippleStateSelected | MDCRippleStateDragged)] =
+        [selectionColor colorWithAlphaComponent:kRippleDraggedAlpha];
+  }
+  self.userInteractionEnabled = YES;
+}
+
+- (UIColor *)rippleColorForState:(MDCRippleState)state {
+  UIColor *rippleColor = _rippleColors[@(state)];
+  if (rippleColor == nil && (state & MDCRippleStateDragged) != 0) {
+    rippleColor = _rippleColors[@(MDCRippleStateDragged)];
+  } else if (rippleColor == nil && (state & MDCRippleStateSelected) != 0) {
+    rippleColor = _rippleColors[@(MDCRippleStateSelected)];
+  }
+
+  if (rippleColor == nil) {
+    rippleColor = _rippleColors[@(MDCRippleStateNormal)];
+  }
+  return rippleColor;
+}
+
+- (void)updateRippleColor {
+  UIColor *rippleColor = [self rippleColorForState:self.state];
+  [self setRippleColor:rippleColor];
+}
+
+- (void)updateActiveRippleColor {
+  UIColor *rippleColor = [self rippleColorForState:self.state];
+  [self setActiveRippleColor:rippleColor];
+}
+
+- (void)setRippleColor:(UIColor *)rippleColor forState:(MDCRippleState)state {
+  _rippleColors[@(state)] = rippleColor;
+
+  [self updateRippleColor];
+}
+
+- (MDCRippleState)state {
+  NSInteger state = 0;
+  if (self.selected) {
+    state |= MDCRippleStateSelected;
+  }
+  if (self.rippleHighlighted) {
+    state |= MDCRippleStateHighlighted;
+  }
+  if (self.dragged) {
+    state |= MDCRippleStateDragged;
+  }
+  return state;
+}
+
+- (void)setSelected:(BOOL)selected {
+  if (!self.allowsSelection || _tapWentOutsideOfBounds) {
+    // If we disallow selection we don't want to apply any visual or state changes for selection.
+    return;
+  }
+  if (selected == _selected && self.activeRippleLayer) {
+    return;
```

Comment:
What does this early return guard against?

Reply:
Done, explained inline.